---
title: "Metagenomics Next Generation Sequencing Report"
author:
    - "Ion Metagenomics"
date: "`r format(Sys.time(), '%d %B, %Y')`"
params:
   rmd: "report.Rmd"
output:
  html_document:
  highlight: tango
  number_sections: no
  theme: default
  toc: yes
  toc_depth: 3
  toc_float:
    collapsed: no
    smooth_scroll: yes
---

```{r setup, echo=FALSE}

# Set some basic options. You usually do not want your code, messages, 
# warnings etc to show in your actual manuscript however for the first
# run or two these will be set on.
  
knitr::opts_chunk$set(warning=TRUE,
            message=TRUE,
            echo=FALSE,
            cache = TRUE,
            tidy = FALSE, ## remove the auto-formatting
            error=TRUE,
            fig.cap='', dev="png")   

```

```{r load_packages, include=FALSE}
#============================================================
# Load packages (load all the packages here at the beginning)
#============================================================
library(pavian)
library(dplyr)
library(ggplot2)
library(DT)
library(Rsamtools)
library(xfun)
library(htmltools)
library(base64enc)
library(openxlsx)
library(kableExtra)
```

### 1. Sample Summary {.tabset}

```{r get_summary_report}

# one report per sample
report <- read_report(snakemake@input[[1]])
report_summary <- summarize_report(report)
## delete 4th, 5th and 6th columns: artifical_reads, unclassified_reads and microbial_reads
report_summary <- report_summary[,-c(4,5,6)]
report_summary_percent <- report_summary/as.numeric(report_summary[1])
report_summary_percent <- signif(report_summary_percent,4)
report_summary_percent[1,] <- paste0(round(100*report_summary_percent, 2), "%")
report_summary_percent[1] <- report_summary[1]

```

#### Classification summary

```{r}
DT::datatable(report_summary_percent, rownames=FALSE,
             options=list(dom = 't', scrollX = TRUE)
             )
```

#### Raw read numbers

```{r}
DT::datatable(report_summary, rownames=FALSE,
             options=list(dom = 't', scrollX = TRUE)
             )
```

### 2. Classification Results Report

``` {r classification}
raw_classify_results <- read.table(snakemake@input[[3]],header=T,sep="\t",quote="",comment.char="")

## order the table by abundance for final output
order_classify_results <- raw_classify_results[order(-raw_classify_results$abundance),,drop=F]

############################################
## calculate coverage
############################################
# remove column of "numUniqueReads" since it's same as "numReads" column
raw_classify_results <- raw_classify_results[,-6]
order_classify_results <- order_classify_results[,-6]

## global parameters
min_mapped_regions <- snakemake@params[[2]]
mapQ <- snakemake@params[[3]]

##
#' Get the pileup-statistics from a bam file using Rsamtools
#'
#' Returns per-base pileup statistics
#'
#' @param bam_file character(1) or BamFile; BAM file path.
#' @param top_n If not NULL, return at most the pileups of that many sequences. integer(1).
#' @param min_mapq Minimum mapping quality.
#' @param ... Additional arguments for \code{Rsamtools::pileup}
#'
#' @return \code{data.frame} with sequence names, positions, strand and count information
get_pileup <- function(bam_file, min_mapq = 0, top_n = NULL, ...) {

  pileup <- Rsamtools::pileup(bam_file,
                              pileupParam=Rsamtools::PileupParam(max_depth=2500, min_base_quality=0, min_mapq=min_mapq,
                                          min_nucleotide_depth=1, min_minor_allele_depth=0,
                                          distinguish_strands=TRUE, distinguish_nucleotides=FALSE,
                                          ignore_query_Ns=TRUE, include_deletions=TRUE, include_insertions=FALSE,
                                          left_bins=NULL, query_bins=NULL, cycle_bins=NULL)
  )

  if (!is.null(top_n)) {
    best_seqs <-
      utils::tail(sort(tapply(
        pileup$count, pileup$seqnames, sum, na.rm = TRUE
      )), top_n)

    pileup <- subset(pileup, seqnames %in% names(best_seqs))
  }
  pileup <- droplevels(pileup)

  ## disregard nucleotide info - maybe change this in future
  covered_bp <- tapply(pileup$pos,pileup$seqnames,function(x) length(unique(x)))
  sum_count <- tapply(pileup$count,pileup$seqnames,sum)

  attr(pileup,"covered_bp") <- covered_bp
  attr(pileup,"sum_count") <- sum_count
  pileup
}

#' Import binary 'BAM' files into a list structure, 
#' with facilities for selecting what fields and which records are imported, 
#' and other operations to manipulate BAM files

get_bam <- function(bam_file) {
  p2 <- Rsamtools::ScanBamParam(what = c("qname","rname", "strand", "pos", "qwidth", "mapq"))
  res <- as.data.frame(Rsamtools::scanBam(bam_file, param = p2)[[1]])
  res[order(res$rname, res$strand, res$pos), ]
}

#' Get sequence lengths from a bam file
#'
#' @param bam_file character(1) or BamFile; BAM file path.
#'
#' @return named vector with lengths of each sequence.
get_seqlengths <- function(bam_file) {
  bamHeader <- Rsamtools::scanBamHeader(bam_file)
  bamHeader[[1]][["targets"]]
}

#' Get non-overlapping regions mapped by reads for each reference genome

#' first define function that spots overlap
overlap = function(x,y) {
  res = ifelse(x[2] >= y[1],1,0)
  return(res)
}

mapped_regions <- function(x,y){
  dt <- data.frame(start=x, end=y+x-1)
  
  # set starting point (row)
  i = 2
  
  # a loop that compares each row with the previous one and deletes row when it finds overlap
  while(i <= nrow(dt)){
   dt_temp = dt
   if (overlap(dt_temp[i-1,], dt_temp[i,]) == 1) {dt_temp[i,]=NA; dt = dt_temp[complete.cases(dt_temp),]} else {dt = dt_temp; i = i+1}
  }
  nrow(dt)
}

#mapped_regions <- function(x,y){
#    ranges <- IRanges(start = x, end = x + y -1)
#    mapped_regions <- IRanges::reduce(ranges,min.gapwidth=0) # findOverlaps
#    length(mapped_regions)
#}

############################################
## check whether bam file is an empty file
## if yes, no filtered results show including Classification Results Report and Coverage Plot  

info <- file.info(snakemake@input[[2]])

if (info$size == 0){
    classify_results_new <- data.frame()
}else{

############################################
## if no, show Classification Results Report and Coverage Plot

    bam_file <- snakemake@input[[2]]

    bam_pileup <- get_pileup(bam_file,min_mapq=mapQ) ## min_mapq = 0, top_n = NULL, ...
    sum_count <- attr(bam_pileup,"sum_count")
    covered_bp <- attr(bam_pileup,"covered_bp")

    bam <- get_bam(bam_file)
    bam_mapq <- bam %>% dplyr::filter(mapq >= mapQ)
    avg_mapq <- tapply(bam_mapq$mapq, bam_mapq$rname, mean)
    #avg_mapq[is.na(avg_mapq)] <- 0
    avg_mapq <- avg_mapq[!is.na(avg_mapq)]

    seq_lengths <- get_seqlengths(bam_file)
    seq_lengths <- seq_lengths[names(sum_count)]
    
    #nreads <- tapply(bam2_mapq$qname, bam2_mapq$rname, function(x) length(unique(x)))
    #nreads[is.na(nreads)] <- 0
    #nreads <- nreads[!is.na(nreads)]

    #sum_count[setdiff(names(seq_lengths),names(sum_count))] <- 0
    #covered_bp[setdiff(names(seq_lengths),names(covered_bp))] <- 0
    
    # map ids to names
    id2tax <- read.table(snakemake@input[[5]],header=T,sep="\t")
    id2tax <- id2tax[!duplicated(id2tax),]
    row.names(id2tax) <- id2tax[,1]
    id2tax <- id2tax[names(seq_lengths),]
    taxid <- id2tax[names(seq_lengths),2]
    species_names <- id2tax[names(seq_lengths),3]
    genus_names <- id2tax[names(seq_lengths),4]
    family_names <- id2tax[names(seq_lengths),5]
    kingdom_names <- id2tax[names(seq_lengths),6]

    # calculate non-overlapped regions for virus hits
    # first suppose all species hits have 100 non-overlapped regions
    num_regions_all = rep(100,length(seq_lengths))
    names(num_regions_all) = names(seq_lengths)

    virus_hits <- id2tax[id2tax[,6] == "Viruses",1]
    if(length(virus_hits) > 0) {
        virus_bam_mapq <- bam_mapq[bam_mapq$rname %in% virus_hits,]
        num_regions <- virus_bam_mapq %>% dplyr::group_by(rname) %>% dplyr::summarise(mapped_regions(pos,qwidth))
        num_regions <- data.frame(num_regions)
        seqnames <- as.vector(num_regions[,1])
        num_regions <- num_regions[,2]
        num_regions_all[seqnames] = num_regions
    }
    seq_info_df <- data.frame(seqID=names(seq_lengths),
                 taxid=taxid,
                 species=species_names,
                 genus=genus_names,
                 family=family_names,
                 kingdom=kingdom_names,
                 genome_size=seq_lengths,
                 avg_depth=signif(sum_count[names(seq_lengths)]/seq_lengths,3),
                 covered_bp=covered_bp[names(seq_lengths)],
                 #n_reads=nreads[names(seq_lengths)],
                 avg_mapq=signif(avg_mapq[names(seq_lengths)],3),
                 num_regions=num_regions_all[names(seq_lengths)])

    seq_info_df$perc_covered = 100*signif(seq_info_df$covered_bp / seq_info_df$genome_size,3)
    #seq_info_df <- seq_info_df[order(-seq_info_df$n_reads,-seq_info_df$perc_covered), , drop=F]

    ##################################################
    ### identify species detected with two conditions:
    ###   1.  >= cutoff set mapped reads (e.g., 2000), or 
    ###       >= cutoff set abundance (e.g.,1%) 
    ###   2.  >= cutoff set genome covered (e.g.,10%)
    ### by intersecting with seq_info_df
    ##################################################

    #seq_info_df <- seq_info_df[seq_info_df[,2] %in% raw_classify_results[,2], ]
    classfify_results <- raw_classify_results[raw_classify_results[,2] %in% seq_info_df[,2],]
    classify_results_new <- merge(classfify_results,seq_info_df,by.x="taxID",by.y="taxid")

    ## for same taxid with multiple sequences, select the one with the highest percent coverage as a representative
    ### first order the table by perc_covered
    classify_results_new <- classify_results_new[order(-classify_results_new$perc_covered),,drop=F]
    ### then remove duplicate taxids
    classify_results_new <- classify_results_new[!duplicated(classify_results_new[,2]),]

    ## filter virus hits based on min_mapped_regions
    classify_results_virus = classify_results_new[classify_results_new$kingdom == "Viruses" & classify_results_new$num_regions >= min_mapped_regions,]
    classify_results_new = classify_results_new[classify_results_new$kingdom != "Viruses",]
    classify_results_new = rbind(classify_results_new,classify_results_virus)

    ## do not show columns: taxRank,genomeSize,n_reads,species,avg_mapq,taxid,num_regions
    classify_results_new$taxaLineage <- NULL
    classify_results_new$totalScore <- NULL
    classify_results_new$alignedNorm <- NULL
    classify_results_new$taxRank <- NULL
    classify_results_new$genomeSize <- NULL
    #classify_results_new$n_reads <- NULL
    classify_results_new$species <- NULL
    classify_results_new$avg_mapq <- NULL
    classify_results_new$taxid <- NULL
    classify_results_new$num_regions <- NULL

    #classify_results_new <- classify_results_new[classify_results_new$numReads >= 2000 | classify_results_new$abundance >= 0.01,]
    #classify_results_new <- classify_results_new[classify_results_new$perc_covered >= 0.1,]

    classify_results_new <- classify_results_new[order(-classify_results_new$abundance,-classify_results_new$numReads,-classify_results_new$perc_covered),,drop=F]
    #DT::datatable(seq_info_df)

}

```

```{r virus host annotate}

# currently, delete these two columns
order_classify_results$totalScore <- NULL
order_classify_results$alignedNorm <- NULL

# virus whose hosts are plant, amoeba, bacterial, fungal, algal, archaeal or insect
# will not be reported

virus_host = read.table(snakemake@params[[4]],header=T,sep="\t",quote="",comment.char="")
#plant = virus_host[grepl("Viridiplantae",virus_host$host.lineage),] # plants & algal
#amoeba = virus_host[grepl("Tubuline",virus_host$host.lineage),] # amoeba
#bacterial = virus_host[grepl("Bacteria",virus_host$host.lineage),] # bacterial
#fungal1 = virus_host[grepl("Fungi",virus_host$host.lineage),] # fungal
#fungal2 = virus_host[grepl("Fungi",virus_host$host.name),] # fungal
#fungal = rbind(fungal1, fungal2)
#archaeal = virus_host[grepl("Archaea",virus_host$host.lineage),] # archaeal
#algal = virus_host[grepl("Phycodnaviridae",virus_host$virus.lineage),] # also algal
#insect = virus_host[grepl("Insecta",virus_host$host.lineage),] # insect

virus_host_simplified <- virus_host[,c(1,8:12)]
raw_virus_results <- order_classify_results[grepl("k__Viruses",order_classify_results$taxaLineage),]
raw_virus_results <- merge(raw_virus_results,virus_host_simplified,by.x="taxID",by.y="virus.tax.id",all.x=TRUE)

virus_moh = read.table(snakemake@params[[6]],header=T,sep="\t",quote="",comment.char="")
virus_moh_df <- c()
for (i in 1:nrow(raw_virus_results)){
    taxid <- raw_virus_results[i,]$taxID
    matches = virus_moh[grepl(paste0("^",taxid,",","|",",",taxid,",","|",",",taxid,"$"),virus_moh$Descendants),]
    if (nrow(matches) > 0){
        descendants <- strsplit(as.character(matches$Descendants),split=",")
        #if(length(descendants) > 1){ 
            #j <- 2
            #while(j <= length(descendants)){
            #    if( all(is.element(descendants[[j-1]], descendants[[j]])) ){ descendants[j] = NULL; } else { descendants[j-1] = NULL; }
            #}
        #}
        #matches = matches[ matches$Descendants %in% paste(unlist(descendants),collapse=","), ]
        
        # assume that lowest rank has minimal number of descendants
        matches = matches[which(lengths(descendants) == min(lengths(descendants))),]
        
        matches = matches[,c("中文名","分类学地位","危害程度分类","Reference")]
    }
    else{
        matches = c(NA,NA,NA,NA)
        names(matches) = c("中文名","分类学地位","危害程度分类","Reference")
    }
    virus_moh_df <- rbind(virus_moh_df,matches)
}

colnames <- colnames(raw_virus_results)
raw_virus_results <- cbind(raw_virus_results,virus_moh_df)
colnames(raw_virus_results) <- c(colnames,"病原归类","分类学地位","危害程度分类","参考来源")
```

```{r proka/euka pathogenic annotate}
pathogen_info <- read.table(snakemake@params[[5]],header=T,sep="\t",quote="",comment.char="")
pathogen_info <- pathogen_info[grepl("human pathogen",pathogen_info$Keywords,ignore.case=TRUE),]
pathogen_info$Definition <- NULL
pathogen_info$Taxonomy.LINEAGE<- NULL

colnames(pathogen_info) <- c("Taxonomy.ID","KEGG.Pathogen","KEGG.Disease","KEGG.Comment","KEGG.Reference")

cellular_results <- order_classify_results[grepl("Bacteria|Eukaryota",order_classify_results$taxaLineage),]
cellular_results <- cellular_results[!grepl("Chordata",cellular_results),]
cellular_results <- merge(cellular_results,pathogen_info,by.x="taxID",by.y="Taxonomy.ID",all.x=TRUE)

bac_fungi_moh = read.table(snakemake@params[[7]],header=T,sep="\t",quote="",comment.char="")
bac_fungi_moh_df <- c()
for (i in 1:nrow(cellular_results)){
    taxid <- cellular_results[i,]$taxID
    matches = bac_fungi_moh[grepl(paste0("^",taxid,",","|",",",taxid,",","|",",",taxid,"$"),bac_fungi_moh$Descendants),]
    if (nrow(matches) > 0){
        descendants <- strsplit(as.character(matches$Descendants),split=",")
        matches = matches[which(lengths(descendants) == min(lengths(descendants))),]
        matches = matches[,c("中文名","危害程度分类","Reference")]
        if(nrow(matches) > 1){ print(taxid);}
    }
    else{
        matches = c(NA,NA,NA)
        names(matches) = c("中文名","危害程度分类","Reference")
    }
    bac_fungi_moh_df <- rbind(bac_fungi_moh_df,matches)
}

colnames <- colnames(cellular_results)
cellular_results <- cbind(cellular_results,bac_fungi_moh_df)
colnames(cellular_results) <- c(colnames,"病原归类","危害程度分类","参考来源")

```

```{r write into xlsx, message=FALSE}

list_of_datasets <-
  list("raw_classify_results" = order_classify_results, 
       "virus_classify" = raw_virus_results[order(-raw_virus_results$abundance,-raw_virus_results$numReads),,drop=F], 
       "proka-euka_classify" = cellular_results[order(-cellular_results$abundance,-cellular_results$numReads),,drop=F]
      )

fn_classify_results <- gsub(".taxLineage.tsv",".xlsx",snakemake@input[[3]])

write.xlsx(list_of_datasets,fn_classify_results)
```

##### **Viruses**:  

```{r virus_results}

if( any(classify_results_new$kingdom == "Viruses") ){
    output <- classify_results_new[classify_results_new$kingdom == "Viruses",]
    output$kingdom <- NULL
    target_host <- virus_host[virus_host$virus.tax.id %in% output$taxID,]
    target_host = target_host[!grepl("Viridiplantae",target_host$host.lineage),] # plants & algal
    target_host = target_host[!grepl("Tubuline",target_host$host.lineage),] # amoeba
    target_host = target_host[!grepl("Bacteria",target_host$host.lineage),] # bacterial
    target_host = target_host[!grepl("Fungi",target_host$host.lineage),] # fungal
    target_host = target_host[!grepl("Fungi",target_host$host.name),] # fungal
    target_host = target_host[!grepl("Archaea",target_host$host.lineage),] # archaeal
    target_host = target_host[!grepl("Phycodnaviridae",target_host$virus.lineage),] # also algal
    target_host = target_host[!grepl("Insecta",target_host$host.lineage),] # insect
    # if host is unknown, keep the virus
    output1 <- output[! output$taxID %in% virus_host$virus.tax.id, ]
    # if host is known, keep the virus when it has other type of hosts
    output2 <- output[output$taxID %in% target_host$virus.tax.id,]
    virus_output <- rbind(output1,output2)
    if(nrow(virus_output) == 0){
        cat("    Not Detected","\n")
    }else{
        DT::datatable(virus_output,rownames = FALSE, ## extensions = 'Buttons',
              options = list(dom = 't', scrollX = TRUE)
              )
    }

}else{
    cat("    Not Detected","\n")
}

```

```{r, viral MoH annoate}

if(nrow(virus_output) > 0){
    output_moh <- raw_virus_results[raw_virus_results$taxID %in% virus_output$taxID,]
    if(nrow(output_moh) > 0 ){
        text_formatted <- c("")
        for(i in 1:nrow(output_moh)){
            if(!is.na(output_moh[i,"病原归类"])){
                text = c("##    ",as.character(output_moh[i,'name']),"属于", output_moh[i,"病原归类"], "，分类学地位为",output_moh[i,"分类学地位"],
                         "，危害程度分类为",output_moh[i,"危害程度分类"], "，其参考来源为：")
                text_formatted <- paste0(text_formatted, paste(text_spec(text,color=c(rep("black",3),rep(c("red","black"),3)),
                                 format="html",bold=T), collapse = " "))
                text_formatted <- paste0(text_formatted, text_spec("卫生部2006人间传染的病原微生物名录", format="html", link=output_moh[i,"参考来源"] ))
                text_formatted <- paste0(text_formatted, text_spec("\n",format="html"))
            }
        }
    }
}
```

`r text_formatted`

<br />

##### **Bacteria**:  

```{r bacteria_results}
if( any(classify_results_new$kingdom == "Bacteria") ){
    output <- classify_results_new[classify_results_new$kingdom == "Bacteria",]
    output$kingdom <- NULL
    DT::datatable(output,rownames = FALSE, ## extensions = 'Buttons',
              options = list(dom = 't', scrollX = TRUE)
              )
}else{
    cat("    Not Detected","\n")
}

```

```{r, bacteria MoH annotate}
if( any(classify_results_new$kingdom == "Bacteria") ){
    output_moh <- cellular_results[cellular_results$taxID %in% output$taxID,]
    if(nrow(output_moh) > 0 ){
        text_formatted <- c("")
        for(i in 1:nrow(output_moh)){
            if(!is.na(output_moh[i,"病原归类"])){
                text = c("##    ",as.character(output_moh[i,'name']),"属于", output_moh[i,"病原归类"], 
                         "，危害程度分类为",output_moh[i,"危害程度分类"], "，其参考来源为：")
                text_formatted <- paste0(text_formatted, paste(text_spec(text,color=c(rep("black",3),rep(c("red","black"),2)),
                                 format="html",bold=T), collapse = " "))
                text_formatted <- paste0(text_formatted, text_spec("卫生部2006人间传染的病原微生物名录", format="html", link=output_moh[i,"参考来源"] ))
                text_formatted <- paste0(text_formatted, text_spec("\n",format="html"))
            }
        }
    }
}
```

`r text_formatted`

<br />

#####  **Archaea**:  

```{r archaea_results}
if( any(classify_results_new$kingdom == "Archaea") ){
    output <- classify_results_new[classify_results_new$kingdom == "Archaea",]
    output$kingdom <- NULL
    DT::datatable(output,rownames = FALSE, ## extensions = 'Buttons',
              options = list(dom = 't', scrollX = TRUE)
              )
}else{
    cat("    Not Detected","\n")
}

```

<br />

##### **Fungi/Protozoa**:  

```{r other_results}
if( any(classify_results_new$kingdom == "Eukaryota") ){
    output <- classify_results_new[classify_results_new$kingdom == "Eukaryota",]
    output$kingdom <- NULL
    DT::datatable(output,rownames = FALSE, ## extensions = 'Buttons',
              options = list(dom = 't', scrollX = TRUE)
              )
}else{
    cat("    Not Detected","\n")
}

```

```{r, fungi MoH annotate}
if( any(classify_results_new$kingdom == "Eukaryota") ){
    output_moh <- cellular_results[cellular_results$taxID %in% output$taxID,]
    if(nrow(output_moh) > 0 ){
        text_formatted <- c("")
        for(i in 1:nrow(output_moh)){
            if(!is.na(output_moh[i,"病原归类"])){
                text = c("##    ",as.character(output_moh[i,'name']),"属于", output_moh[i,"病原归类"], 
                         "，危害程度分类为",output_moh[i,"危害程度分类"], "，其参考来源为：")
                text_formatted <- paste0(text_formatted, paste(text_spec(text,color=c(rep("black",3),rep(c("red","black"),2)),
                                 format="html",bold=T), collapse = " "))
                text_formatted <- paste0(text_formatted, text_spec("卫生部2006人间传染的病原微生物名录", format="html", link=output_moh[i,"参考来源"] ))
                text_formatted <- paste0(text_formatted, text_spec("\n",format="html"))
            }
        }
    }
}
```

`r text_formatted`

*Note: You could also see section 6 for more possible results.*

### 4. Antibiotic Resistance Genes Results 

```{r}
arg_table = read.table(snakemake@input[[7]],header=T,sep="\t",quote="",comment.char="") 

if(nrow(arg_table) == 0){
    cat("    Not Detected","\n")
}else{
    arg <- arg_table[arg_table[,'perc_covered'] >= 0.85 & arg_table[,'perc_identity'] >= 0.95],]
    if(nrow(arg) > 0){
        DT::datatable(arg,rownames = FALSE, ## extensions = 'Buttons',
              options = list(dom = 't', scrollX = TRUE)
              )
    }else{
        cat("    Not Detected","\n")
    }
}

#### Predicted Antibiotic Resistance Drug Class PieChart
# Currently, do not show this figure
# if(nrow(arg_table) == 0){
#    cat("    Not Detected","\n")
#}else{
#    knitr::include_graphics(snakemake@input[[8]])
#}

```

### 5. Coverage View

```{r plot_coverage,fig.width=10,fig.height=3.5}

# plot genome coverage

if( nrow(virus_output) > 0 ){
    cat("Detected Virus Coverage Plot","\n")
    for( name in virus_output$seqID ){
        plot_pileup_act <- plot_pileup(bam_pileup[bam_pileup[,1] == name,])
        ## in a loop, use print to visibly display the plot
        print( plot_pileup_act + ggtitle(paste0(name," | ", id2tax[id2tax[,1] == name,3])) + theme(plot.title = element_text(hjust = 0.5)) )
    }
}

if( any(classify_results_new$kingdom == "Bacteria") ){
    output <- classify_results_new[classify_results_new$kingdom == "Bacteria",]
    cat("Detected Bacteria Coverage Plot","\n")
    for( name in output$seqID ){
        plot_pileup_act <- plot_pileup(bam_pileup[bam_pileup[,1] == name,])
        ## in a loop, use print to visibly display the plot
        print( plot_pileup_act + ggtitle(paste0(name," | ", id2tax[id2tax[,1] == name,3])) + theme(plot.title = element_text(hjust = 0.5)) )
    }
}

if( any(classify_results_new$kingdom == "Archaea") ){
    output <- classify_results_new[classify_results_new$kingdom == "Archaea",]
    cat("Detected Archaea Coverage Plot","\n")
    for( name in output$seqID ){
        plot_pileup_act <- plot_pileup(bam_pileup[bam_pileup[,1] == name,])
        ## in a loop, use print to visibly display the plot
        print( plot_pileup_act + ggtitle(paste0(name," | ", id2tax[id2tax[,1] == name,3])) + theme(plot.title = element_text(hjust = 0.5)) )
    }
}

if( any(classify_results_new$kingdom == "Eukaryota") ){
    output <- classify_results_new[classify_results_new$kingdom == "Eukaryota",]
    cat("Detected Fungi/Parasite Coverage Plot","\n")
    for( name in output$seqID ){
        plot_pileup_act <- plot_pileup(bam_pileup[bam_pileup[,1] == name,])
        ## in a loop, use print to visibly display the plot
        print( plot_pileup_act + ggtitle(paste0(name," | ", id2tax[id2tax[,1] == name,3])) + theme(plot.title = element_text(hjust = 0.5)) )
    }
}


    #plot_pileup_act <- plot_pileup(bam_pileup[bam_pileup[,1] == name,])
    #plot_pileup_act <- plot_pileup(bam_pileup[bam_pileup[,1] == name,], seq_info_df[name,], text_size = 4)
    
    ## in a loop, use print to visibly display the plot
    #print( plot_pileup_act + ggtitle(paste0(name," | ", id2tax[id2tax[,1] == name,3])) + theme(plot.title = element_text(hjust = 0.5)) )

    #ggsave(paste0(snakemake@params[[1]],name,".cov.pdf"), plot_pileup_act + 
    #                          ggtitle(name) + theme(plot.title = element_text(hjust = 0.5)),
    #                          "pdf", height = 2.5, units = "in")

```

### 6. Raw Classification Results

```{r download_raw_data}
    #' define an embed_files function: first zip files, and then embed it
    
    embed_files = function(path, name = with_ext(gsub("_se.report.xlsx","",basename(path[1])), '.RawClassifyResults.zip'), ...) {
      name = file.path(tempdir(), basename(name))
      on.exit(file.remove(name), add = TRUE)
      utils::zip(name, path, zip="zip",flags="-rj9X")
      xfun::embed_file(name, ...)
    }
    embed_files(c(fn_classify_results,snakemake@input[[4]]))
```

### 7. Viral Assembly Results

```{r assembly_results}

contigs_folder <- gsub("assembly.stat.txt","",snakemake@input[[6]])
if( nrow(virus_output) > 0 ){
    for( taxid in virus_output$taxID ){
        contigs_file = paste0(contigs_folder,taxid, "/", paste0(taxid,".contigs.fa"))
        contigs_list <- c()
        info <- file.info(contigs_file)
        if (info$size > 0){
            contigs_list <- c(contigs_list,contigs_file)
        }
    }
    if(length(contigs_list) > 0){
        embed_files(contigs_list,name = with_ext(gsub("_se.report.xlsx","",basename(fn_classify_results)), '.ViralAssemblyResults.zip'))
    }else{
        cat("    No Assembly Results Available", "\n")
    }
   
}
```

### About
This is for research use only. It is not intended for any animal or human therapeutic or diagnostic use. 


### Source
<a download="report.Rmd" href="`r base64enc::dataURI(file = params$rmd, mime = 'text/rmd', encoding = 'base64')`">R Markdown source file (to produce this document)</a>